File: 250704/거꾸로 순열/backward-permutation.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');
const n = Number(input[0]);
const visited = Array(n+1).fill(false)
const picked = [];
function permutation(cnt){

    if(cnt === n){
        console.log(picked.join(' '))
    }

    for(let i = n; i >= 1; i--){
        if(visited[i] === true) continue;

        visited[i] = true;
        picked.push(i);

        permutation(cnt+1);

        visited[i] = false;
        picked.pop() 
    }
}

permutation(0)

File: 260123/거꾸로 순열/backward-permutation.py
n = int(input())
visited = [False] * (n + 1)
picked = []

def dfs(cnt):
    if cnt == n:
        for v in picked:
            print(v, end=' ')
        print()
        return
    
    for i in range(n, 0, -1):
        if visited[i]:
            continue

        visited[i] = True
        picked.append(i)

        dfs(cnt + 1)

        visited[i] = False
        picked.pop()
dfs(0)